File: decoratedstr.py
# -*- encoding: utf-8 -*-

from __future__ import print_function, unicode_literals
import six

__all__ = ['remove_decoration', 'decorated_match']

# for range \u00c0 \u0179
char_to_alternatives_lower = {
    'a': 'àáâãäåāăą',
    'c': 'çćĉċč',
    'd': 'ďđ',
    'e': 'èéêëēĕėęě',
    'g': 'ĝğġģ',
    'h': 'ĥħ',
    'i': 'ìíîïĩīĭįı',
    'j': 'ĵ',
    'k': 'ķ',
    'l': 'ĺļľŀł',
    'n': 'ñńņňŉŋ',
    'o': 'òóôöøōŏő',
    'r': 'ŕŗř',
    's': 'śŝşš',
    't': 'ţťŧ',
    'u': 'ùúûüũūŭůűų',
    'w': 'ŵ',
    'y': 'ýÿŷ',
    'z': 'źżž',
}

# This chars lower() function doesn't work
char_to_alternatives_upper = {
    'I': 'İ',
}
char_to_alternatives = {} # idem, but with upper case too
for char, alternatives in six.iteritems(char_to_alternatives_lower):
    char_to_alternatives[char] = alternatives
    char_to_alternatives[char.upper()] = alternatives.upper()
for char, alternatives in six.iteritems(char_to_alternatives_upper):
    char_to_alternatives[char] = alternatives

alternative_to_char = {} # reverse
for char, alternatives in six.iteritems(char_to_alternatives_lower):
    for alternative in alternatives:
        alternative_to_char[alternative] = char
        alternative_to_char[alternative.upper()] = char.upper()
for char, alternatives in six.iteritems(char_to_alternatives_upper):
    for alternative in alternatives:
        alternative_to_char[alternative] = char

# ligatures (only two chars supported)
ligatures_expansions_lower = {
    'æ': 'ae',
    # 'ĳ': 'ij', buggy: see http://en.wikipedia.org/wiki/Typographic_ligature
    'œ': 'oe',
}

ligatures_expansions = {} # idem, but with upper case too
for ligature, expansion in six.iteritems(ligatures_expansions_lower):
    ligatures_expansions[ligature] = expansion
    ligatures_expansions[ligature.upper()] = expansion[0].upper()+expansion[1:]

ligatures_contractions = {} # reverse
for ligature, expansion in six.iteritems(ligatures_expansions_lower):
    ligatures_contractions[expansion] = ligature
    ligatures_contractions[expansion[0].upper()+expansion[1:]] = ligature.upper()

def remove_decoration(txt):
    result = ''
    for l in txt:
        l = alternative_to_char.get(l, l)
        l = ligatures_expansions.get(l, l)
        result += l
    return result

def decorated_match_single_char(c, casesensitive=False):
    assert type(c) == type('')
    if not casesensitive:
        c = c.lower()
    result = c + char_to_alternatives.get(c, '')
    if not casesensitive:
        u = result.upper()
        if result != u:
            result += u
    if len(result) > 1:
        return '['+result+']'
    else:
        return result

def decorated_match(txt, casesensitive=False):
    assert type(txt) == type('')
    result = ''
    txt = remove_decoration(txt)
    if not casesensitive:
        txt = txt.lower()
    i = 0
    while i < len(txt):
        c1 = txt[i] # current character
        c12 = txt[i:i+2] # both current and next characters. Contains a single char on last iteration so that it never matches, that is OK
        ligature = ligatures_contractions.get(c12, None)
        if ligature:
            result += '('+ligature
            if not casesensitive:
                result += '|'+ligature.upper()
            result += '|'+decorated_match_single_char(c12[0], casesensitive) \
                          +decorated_match_single_char(c12[1], casesensitive) \
                   +')'
            i += 1 # skip next character, we allready did both
        else:
            result += decorated_match_single_char(c1, casesensitive)
        i += 1
    return result


if __name__ == '__main__':
    import sys
    from optparse import OptionParser
    parser = OptionParser(usage='%prog [options] string')
    parser.add_option('--charset', help="set charset. default=%default", action='store', dest='charset', default='utf-8')
    parser.add_option('-r', '--regexp', help="generate regular expression.", action='store_true', dest='regexp')
    parser.add_option('-i', help="used with -r, make regexp case insensitive.", action='store_false', dest='casesensitive', default=True)
    (options, args) = parser.parse_args()

    if not args:
        print('Missing required parameter. Try "Œuf"', file=sys.stderr)
        sys.exit(1)
    if six.PY3:
        input = ' '.join(args)
    else:
        input = unicode(b' '.join(args), options.charset)
    #print("input:", input)                            # Œuf
    #print("undecorated:", remove_decoration(input))   # Oeuf
    #print("regex:", decorated_match(input))           # (œ|Œ|[oòóôöøōŏőOÒÓÔÖØŌŎŐ][eèéêëēĕėęěEÈÉÊËĒĔĖĘĚ])[uùúûüũūŭůűųUÙÚÛÜŨŪŬŮŰŲ][fF]
    if options.regexp:
        if six.PY3:
            print(decorated_match(input, options.casesensitive))
        else:
            print(decorated_match(input, options.casesensitive).encode(options.charset))
    else:
        if six.PY3:
            print(remove_decoration(input))
        else:
            print(remove_decoration(input).encode(options.charset))

File: DecoratedStr.java
import java.util.Hashtable;
import java.util.Enumeration;
public class DecoratedStr {
    protected static Hashtable<Character,String> char_to_alternatives_lower = new Hashtable<Character,String>();
    protected static Hashtable<Character,String> char_to_alternatives = new Hashtable<Character,String>();
    protected static Hashtable<Character,Character> alternative_to_char = new Hashtable<Character,Character>();
    protected static Hashtable<Character,String> ligatures_expansions_lower = new Hashtable<Character,String>();
    protected static Hashtable<Character,String> ligatures_expansions = new Hashtable<Character,String>();
    protected static Hashtable<String,Character> ligatures_contractions = new Hashtable<String,Character>();
    private static boolean initialized = false;

    public static void init() {
        if (initialized)
            return;
        char_to_alternatives_lower.put('a', "àáâãäåāăą");
        char_to_alternatives_lower.put('c', "çćĉċč");
        char_to_alternatives_lower.put('d', "ďđ");
        char_to_alternatives_lower.put('e', "èéêëēĕėęě");
        char_to_alternatives_lower.put('g', "ĝğġģ");
        char_to_alternatives_lower.put('h', "ĥħ");
        char_to_alternatives_lower.put('i', "ìíîïĩīĭįı");
        char_to_alternatives_lower.put('j', "ĵ");
        char_to_alternatives_lower.put('k', "ķ");
        char_to_alternatives_lower.put('l', "ĺļľŀł");
        char_to_alternatives_lower.put('n', "ñńņňŉŋ");
        char_to_alternatives_lower.put('o', "òóôöøōŏő");
        char_to_alternatives_lower.put('r', "ŕŗř");
        char_to_alternatives_lower.put('s', "śŝşš");
        char_to_alternatives_lower.put('t', "ţťŧ");
        char_to_alternatives_lower.put('u', "ùúûüũūŭůűų");
        char_to_alternatives_lower.put('w', "ŵ");
        char_to_alternatives_lower.put('y', "ýÿŷ");
        char_to_alternatives_lower.put('z', "źżž");

        Enumeration<Character> e = char_to_alternatives_lower.keys();
        while (e.hasMoreElements()) {
            Character k = e.nextElement();
            String v = char_to_alternatives_lower.get(k);
            char_to_alternatives.put(k, v);
            char_to_alternatives.put(Character.toUpperCase(k), v.toUpperCase());
            //System.out.println(Character.toUpperCase(k));
            //System.out.println(v.toUpperCase());
        }
        char_to_alternatives.put('I', "İ");

        e = char_to_alternatives.keys();
        while (e.hasMoreElements()) {
            Character k = e.nextElement();
            String v = char_to_alternatives.get(k);
            for (int i=v.length()-1; i>=0; --i) {
                char a = v.charAt(i);
                //System.out.println(k);
                //System.out.println(a);
                alternative_to_char.put(a, k);
            }
        }
        
        ligatures_expansions_lower.put('æ', "ae");
        //ligatures_expansions_lower.put('ĳ', "ij"); buggy: see http://en.wikipedia.org/wiki/Typographic_ligature
        ligatures_expansions_lower.put('œ', "oe");

        e = ligatures_expansions_lower.keys();
        while (e.hasMoreElements()) {
            Character k = e.nextElement();
            String v = ligatures_expansions_lower.get(k);
            ligatures_expansions.put(k, v);
            ligatures_contractions.put(v, k);
            String uv = Character.toUpperCase(v.charAt(0)) + v.substring(1);
            ligatures_expansions.put(Character.toUpperCase(k), uv);
            ligatures_contractions.put(uv, Character.toUpperCase(k));
        }

        initialized = true;
    }

    public static String remove_decoration(String txt) {
        init();
        String result = "";
        int len = txt.length();
        char l;
        Character al;
        String le;
        for (int i=0; i<len; ++i) {
            l = txt.charAt(i);
            al = alternative_to_char.get(l);
            if (al != null)
                l = al;
            le = ligatures_expansions.get(l);
            if (le != null)
                result += le;
            else
                result += l;
        }
        return result;
    }

    public static String decorated_match_single_char(char c, boolean case_sensitive) {
        init();
        if (!case_sensitive)
            c = Character.toLowerCase(c);
        String result = "" + c;
        String sa = char_to_alternatives.get(c);
        if (sa != null)
            result += sa;
        if (!case_sensitive) {
            String ur = result.toUpperCase();
            if (result.compareTo(ur) != 0)
                result += ur;
        }
        if (result.length() > 1)
            return "[" + result + "]";
        else
            return result;

    }

    public static String decorated_match(String txt, boolean case_sensitive) {
        init();
        String result = "";
        txt = remove_decoration(txt);
        if (!case_sensitive)
            txt = txt.toLowerCase();
        int len = txt.length();
        for (int i=0; i<len; ++i) {
            char c1 = txt.charAt(i);
            String c12 = "" + c1;
            if (i < len-1)
                c12 += txt.charAt(i+1);
            Character lc = ligatures_contractions.get(c12);
            if (lc != null) {
                result += "(" + lc;
                if (!case_sensitive)
                    result += "|" + Character.toUpperCase(lc);
                result += '|'
                       + decorated_match_single_char(c12.charAt(0), case_sensitive)
                       + decorated_match_single_char(c12.charAt(1), case_sensitive)
                       + ')';
                i += 1;

            }
            else
                result += decorated_match_single_char(c1, case_sensitive);
        }
        return result;
    }

	public static void main(String argv[]) {
        String in = "Œuf";
		System.out.println(in);
		System.out.println(remove_decoration(in));
		//System.out.println(decorated_match_single_char('m', true));
		//System.out.println(decorated_match_single_char('m', false));
		//System.out.println(decorated_match_single_char('h', true));
		//System.out.println(decorated_match_single_char('h', false));
		System.out.println(decorated_match(in, true));
		System.out.println(decorated_match(in, false));
	}
}
